File: xml_generator/KDG.py
# classes for KDG


class Node:
    """a node
    """

    def __init__(self, node_name, cost, quality):
        """ Initialization
        :param node_name: name of node
        :param cost: cost of the node
        :param quality: quality of the node
        """
        self.name = node_name
        self.cost = cost
        self.quality = quality


class Knob:
    """ a collection of Node, referenced by lvl (id)
    """

    def __init__(self, knob_name):
        self.nodes = {}
        self.name = knob_name

    def addNode(self, node, lvl):
        """ Add a node
        :param node: the node to be added
        """
        self.nodes[lvl] = node

    def getNode(self, lvl):
        """ Get a node
        :param lvl: level of the node
        """
        return self.nodes[lvl]


class Constraint:
    """ a node pairwise constraint
    The semantic of a constraint in a KDG graph is:
    sink <----- source
    if source_node is chosen, then the sink_node must be chosen
    """

    def __init__(self, source_node, sink_node):
        self.source_node = source_node
        self.sink_node = sink_node


class KDG:
    """ a collection of a knobs and a list of constraints
    """

    def __init__(self):
        self.knobs = {}
        self.constraints = []

    def setName(self, app_name):
        self.name = app_name

    def addKnob(self, knob):
        self.knobs[knob.name] = knob

    def addConstraint(self, constraint):
        self.constraints.append(constraint)

    def addConstraints(self, constraints):
        self.constraints.extend(constraints)

    def getConstraintForNode(self, node_name):
        """ retrieve the list of Constraints in which the node_name is the sink
        :param node_name: the name of the node (sink)
        """
        sources = []
        for c in self.constraints:
            if c.sink_node == node_name:
                sources.append(c)
        return sources

File: lp_generator/KDG.cpp
#include "KDG.h"
#include <iostream>

using namespace std;

/****** Node ******/

void Node::setName(string n) { name_ = n; }

string Node::getName() { return name_; }


/****** Basic ******/

Basic::Basic(string basicName):cost_(0.0), quality_(0.0) {
    setName(basicName);
}

void Basic::setCost(float val) { cost_ = val; }

void Basic::setQuality(float val) { quality_ = val; }

float Basic::getCost() { return cost_; }

float Basic::getQuality() { return quality_; }

void Basic::addDependency(Basic * source) { dependencies_.push_back(source); }

/****** Level ******/


Level::Level(int lvl):level_(lvl) {}

vector<Basic *> *Level::getBasicNodes() { return &basicNodes_; }

void Level::addBasicNode(Basic *basic) { basicNodes_.push_back(basic); }

int Level::getLevelNum() { return level_; }

Level::~Level() {
    for (Basic *basic : basicNodes_) {
        delete basic;
    }
}


/****** Knob ******/

Knob::Knob(string knob_name){
    setName(knob_name);
}

vector<Level *> *Knob::getLevelNodes() { return &levelNodes_; }

void Knob::addLevelNode(Level *level) { levelNodes_.push_back(level); }

Knob::~Knob() {
    for (Level *level : levelNodes_) {
        delete level;
    }
}


/****** KDG ******/

KDG::KDG(string appName):appName_(appName) {}

void KDG::addKnob(Knob *knob) {
    knobs_.push_back(knob);
}

Node *KDG::getNodeFromName(string name) {
    for (Knob *knob : knobs_) {
        if (knob->getName().compare(name) == 0) {
            return knob;
        }
        for (Level *lvl : *(knob->getLevelNodes())) {
            for (Basic *b : *(lvl->getBasicNodes())) {
                if (b->getName().compare(name) == 0) {
                    return b;
                }
            }
        }
    }
    return NULL;
}

string KDG::getName(){
    return appName_;
}

KDG::~KDG() {
    for (Knob *knob : knobs_) {
        delete knob;
    }
}
